### PR TITLE
Post message current page's script's URL to SW for caching

### PR DESCRIPTION
**Pull request by prateekbh**
As a part of Auto-PWA effort AMP will build a lazy service worker for publishers. Being lazy, these service workers will have no access to what resources are used to cache for first visit.

- This change will use see if there is a new serviceworker being installed
- If yes, it'll wait for that serviceworker to be active
- use ResourceTiming API to see what AMP scripts were used in the current page and postMessage them to service worker to add them to SW cache.

### CODE REVIEW

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 1]
```diff
@@ -296,12 +296,40 @@ function install(win, src) {
       user().info(TAG, 'ServiceWorker registration successful with scope: ',
           registration.scope);
     }
+    sendScriptAndURLToSWOnFirstVisit(registration, win);
```
Comment (jridgewell) [inline extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: `Url`, `Sw`.

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 2]
```diff
@@ -296,12 +296,40 @@ function install(win, src) {
...
+ * @param {ServiceWorkerRegistration} registration
+ * @param {!Window} win
+ */
+function sendScriptAndURLToSWOnFirstVisit(registration, win) {
+  const installingServiceWorker = registration.installing;
+  if (installingServiceWorker && 'performance' in win) {
+    installingServiceWorker.addEventListener('statechange', evt => {
+      if (evt.target.state === 'activated' &&
+        win.navigator.serviceWorker.controller) {
+        // Fetch all AMP-scripts used on the page
```
Comment (jridgewell) [inline extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: Nit: let's early return and unnest.

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js [hunk 3]
```diff
@@ -296,12 +296,40 @@ function install(win, src) {
...
+        const ampScriptsUsed = win.performance.getEntriesByType('resource')
+            .filter(item => item.initiatorType === 'script' &&
+                /https:\/\/cdn.ampproject.org\//.test(item.name))
+            .map(script => script.name);
+        // using https://github.com/redux-utilities/flux-standard-action
+        win.navigator.serviceWorker.controller.postMessage(JSON.stringify({
```
Comment (jridgewell) [inline extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: Nit: store `win.navigator.serviceWorker` into a variable.

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js (hunk 2 above)
Comment (prateekbh) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: done

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js (hunk 3 above)
Comment (prateekbh) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: done

File: extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js (hunk 1 above)
Comment (prateekbh) [reply extensions/amp-install-serviceworker/0.1/amp-install-serviceworker.js]: done

File: build-system/tasks/presubmit-checks.js [hunk 4]
```diff
@@ -700,14 +700,19 @@ const forbiddenTermsSrcInclusive = {
...
   '\\.getMatchedCSSRules\\(': bannedTermsHelpString,
-  '\\.postMessage\\(': bannedTermsHelpString,
   '\\.scrollBy\\(': bannedTermsHelpString,
   '\\.scrollIntoView\\(': bannedTermsHelpString,
   '\\.scrollIntoViewIfNeeded\\(': bannedTermsHelpString,
   '\\.scrollTo\\(': bannedTermsHelpString,
   '\\.webkitConvertPointFromNodeToPage\\(': bannedTermsHelpString,
   '\\.webkitConvertPointFromPageToNode\\(': bannedTermsHelpString,
   '\\.scheduleUnlayout\\(': bannedTermsHelpString,
+  '\\.postMessage\\(': {
```
Comment (prateekbh) [inline build-system/tasks/presubmit-checks.js]: @jridgewell I added this as `amp-install-serviceworker` need to postMessage, plz review

Comment (jridgewell) [reply build-system/tasks/presubmit-checks.js]: I'd be surprised if this whitelist is complete. But a good change.

Comment (prateekbh) [reply build-system/tasks/presubmit-checks.js]: wouldn't throw an error if this wasn't complete?

Comment (prateekbh) [reply build-system/tasks/presubmit-checks.js]: Ah with `ok`s....
I'll open another PR may be to clean them

**Pull request merged**